File: trunk/Geometry/Edge/length.cpp
#include "length.h"

#include <stdio.h>

typedef map<TriPosition, Length*, TriPositionCompare> LengthIndex;
static LengthIndex* Index = NULL;

Length::Length( Edge& e ){
  Vertex v1 = Triangulation::vertexTable[ (*(e.getLocalVertices()))[0] ];
  Vertex v2 = Triangulation::vertexTable[ (*(e.getLocalVertices()))[1] ];

  radius1 = Radius::At( v1 );
  radius2 = Radius::At( v2 );
  eta = Eta::At( e );
  
  radius1->addDependent( this );
  radius2->addDependent( this );
  eta->addDependent( this );
}

void Length::recalculate(){
  double r1 = radius1->getValue();
  double r2 = radius2->getValue();
  double etaV = eta->getValue();

  value = sqrt( r1*r1 + r2*r2 + 2*r1*r2*etaV );   
}

void Length::remove() {
   deleteDependents();
   radius1->removeDependent(this);
   radius2->removeDependent(this);
   eta->removeDependent(this);
   Index->erase(pos);
   delete this;
}

Length::~Length(){}

Length* Length::At( Edge& e ){
  TriPosition T( 1, e.getSerialNumber() );
  if( Index == NULL ) Index = new LengthIndex();
  LengthIndex::iterator iter = Index->find( T );

  if( iter == Index->end() ){
    Length* val = new Length( e );
    val->pos = T;
    Index->insert( make_pair( T, val ) );
    return val;
  } else {
    return iter->second;
  }
}

void Length::CleanUp(){
  if( Index == NULL ) return;
  LengthIndex::iterator iter;
  LengthIndex copy = *Index;
  for(iter = copy.begin(); iter != copy.end(); iter++) {
    iter->second->remove();
  }
    
  delete Index;
  Index = NULL;
}

void Length::print(FILE* out) {
  map<int, Edge>::iterator eit;
  
  fprintf(out, "Length [ e ]\n=============\n");
  for(eit = Triangulation::edgeTable.begin(); eit != Triangulation::edgeTable.end(); eit++) {
     fprintf(out, "Length [%3d]\t= % 2.8f\n", eit->first, Length::valueAt(eit->second));
  }       
}

File: SKUNK/src/Geoquant/Length.java
package Geoquant;

import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Set;

import Triangulation.*;

public class Length extends Geoquant {
  // Index map
  private static HashMap<TriPosition, Length> Index = new HashMap<TriPosition, Length>();
  
  // Needed geoquants
  private Radius rad1;
  private Radius rad2;
  private Alpha alpha1;
  private Alpha alpha2;
  private Eta eta;
  
  public Length(Edge e) {
    super();
    
    List<Vertex> verts = new ArrayList<Vertex>();
    verts.addAll(e.getLocalVertices());
    
    rad1 = Radius.At(verts.get(0));
    rad1.addDependent(this);
    alpha1 = Alpha.At(verts.get(0));
    alpha1.addDependent(this);
    rad2 = Radius.At(verts.get(1));
    rad2.addDependent(this);
    alpha2 = Alpha.At(verts.get(1));
    alpha2.addDependent(this);
    eta = Eta.At(e);
    eta.addDependent(this);
  }
  
  protected void recalculate() {
    double r1 = rad1.getValue();
    double r2 = rad2.getValue();
    double a1 = alpha1.getValue();
    double a2 = alpha2.getValue();
    double etaV = eta.getValue();

    value = Math.sqrt( a1* r1*r1 + a2 * r2*r2 + 2*r1*r2*etaV );

  }
 
  protected void remove() {
    deleteDependents();
    rad1.removeDependent(this);
    rad2.removeDependent(this);
    alpha1.removeDependent(this);
    alpha2.removeDependent(this);
    eta.removeDependent(this);
    Index.remove(pos);
  }
  
  public static Length At(Edge e) {
    TriPosition T = new TriPosition(e.getSerialNumber());
    Length q = Index.get(T);
    if(q == null) {
      q = new Length(e);
      q.pos = T;
      Index.put(T, q);
    }
    return q;
  }
  
  public static double valueAt(Edge e) {
    return At(e).getValue();
  }

}
